src/runtime/relax_vm/paged_kv_cache.cc

```diff
@@ -0,0 +1,633 @@
+    for (int64_t page_idx = cur_npage; page_idx < tgt_npage; ++page_idx) {
+      AllocatePageForSequence(seq_id);
+    }
+    seq_lengths_[seq_id] += extra_length;
+    cur_append_lengths_[seq_id] = extra_length;
+    dirty_aux_data_device_ = true;
+  }
+
+  /*!
+   * \brief The entrance of attention. It takes an attention compute
+   * function together with the query data, invokes the attention
+   * function to complete the computation.
+   * \param f_attention The input attention compute function.
+   * \param q_data The query data. We support the following two layout settings:
+   * - in batch decode settings, q_data has layout (num_total_seqs, 1, num_heads, head_dim)
+   *   where num_total_seqs should exactly equal to the number of sequences in the cache.
+   * - in other settings (single-sequence prefill, batch prefill, or speculation
+   *   verification), q_data has the **flattened layout** to handle raggedness:
+   *   (1, total query length, num_heads, head_dim).
+   * \param layer_id The model layer index of the current attention.
+   * \param output The attention output array.
+   * \param apply_rotary A boolean flag indicating if to apply RoPE.
+   * \param rotary_scale The RoPE scale if applicable.
+   * \param rotary_theta The RoPE theta if applicable.
+   */
+  void Attention(PackedFunc f_attention, NDArray q_data, int64_t layer_id, NDArray output,
+                 bool apply_rotary = false, double rotary_scale = 1.0f, double rotary_theta = 1e4) {
+    // Check q_data shape validity.
+    CHECK_EQ(q_data->ndim, 4);
+    CHECK_GT(q_data->shape[1], 0);
+    CHECK_EQ(q_data->shape[2], num_heads_);
+    CHECK_EQ(q_data->shape[3], head_dim_);
+    CHECK(q_data.DataType() == pages_.DataType());
+
+    if (q_data->shape[0] > 1) {
+      CHECK_EQ(q_data->shape[0], num_total_seqs_);
+      CHECK_EQ(q_data->shape[1], 1);
+    }
+    int64_t ntoken = 0;
+    for (int64_t seq_id = 0; seq_id < num_total_seqs_; ++seq_id) {
+      ntoken += cur_append_lengths_[seq_id];
+      CHECK_LE(cur_append_lengths_[seq_id], seq_lengths_[seq_id]);
+      if (q_data->shape[0] > 1) {
+        CHECK_EQ(cur_append_lengths_[seq_id], 1);
+      }
+    }
+    CHECK_EQ(ntoken, q_data->shape[0] * q_data->shape[1]);
+
+    // The auxiliary data structure on device must have been synchronized.
+    CHECK(!dirty_aux_data_device_)
+        << "The auxiliary arrays are not synchronized to device. Please call "
+           "`SyncAuxArrayToDevice` to synchronize before calling `Attention`.";
+
+    // Provide a 8MB size float temp buffer for the attention kernel.
+    static NDArray tmp_buffer =
+        NDArray::Empty({8 * 1024 * 1024}, DLDataType(DataType::Float(32)), pages_->device);
+
+    f_attention(q_data, pages_,                                                                  //
+                page_table_indptr_device_.CreateView({num_total_seqs_ + 1}, dtype_aux_),         //
```

Comment:
would be great these views are also created once during sync device and kept there without having to recreate on the fly